Source/Session.swift

```diff
@@ -552,69 +577,60 @@ extension Session: RequestDelegate {
         }
     }
 
-    public func cancelRequest(_ request: Request) {
-        rootQueue.async {
-            request.didCancel()
+    public func attemptToTransition(_ request: Request, to state: Request.State) {
+        requestStateQueue.sync {
+            guard request.attemptToTransitionTo(state) else { return }
 
-            // Cancellation only has an effect on running or suspended tasks.
-            guard let task = self.requestTaskMap[request], [.running, .suspended].contains(task.state) else {
-                request.finish()
-                return
-            }
+            switch state {
+            case .resumed:
+                request.didResume()
+
+                guard let task = request.task else { return }
```

Comment:
By using the `request.task` instead of the `requestTaskMap`, we avoid having to lock the `RequestTaskMap` since it could be accessed from different threads.